### PR TITLE
Added Native-specific frontend checker for `@SharedImmutable`

### PR DESCRIPTION
**Pull request by LepilkinaElena**

### CODE REVIEW

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+object NativeSharedImmutableChecker : DeclarationChecker {
+    private val sharedImmutableFqName = FqName("kotlin.native.concurrent.SharedImmutable")
+
+    override fun check(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (descriptor !is VariableDescriptor || !descriptor.isVar) return
+        val sharedImmutableAnnotation = descriptor.annotations.findAnnotation(sharedImmutableFqName)
+        sharedImmutableAnnotation?.let {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(it) ?: declaration
+            context.trace.report(ErrorsNative.INCOMPATIBLE_SHARED_IMMUTABLE.on(reportLocation))
+        }
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt]: Maybe also report if `@SharedImmutable`-annotated declaration is not top-level?

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt [hunk 2]
```diff
@@ -20,6 +20,7 @@ private val DIAGNOSTIC_FACTORY_TO_RENDERER by lazy {
             ErrorsNative.INCOMPATIBLE_THROWS_INHERITED, "Member inherits different @Throws filters from {0}",
             Renderers.commaSeparated(Renderers.NAME)
         )
+        put(ErrorsNative.INCOMPATIBLE_SHARED_IMMUTABLE, "@SharedImmutable isn't applicable to mutable variable")
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt]: > variable

property (or field) actually.

Comment (SvyatoslavScherbina) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt]: Or "@SharedImmutable isn't applicable to var".

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/ErrorsNative.kt [hunk 3]
```diff
@@ -20,6 +20,8 @@ object ErrorsNative {
     val INCOMPATIBLE_THROWS_OVERRIDE = DiagnosticFactory1.create<KtElement, DeclarationDescriptor>(Severity.ERROR)
     @JvmField
     val INCOMPATIBLE_THROWS_INHERITED = DiagnosticFactory1.create<KtDeclaration, Collection<DeclarationDescriptor>>(Severity.ERROR)
+    @JvmField
+    val INCOMPATIBLE_SHARED_IMMUTABLE = DiagnosticFactory0.create<KtElement>(Severity.ERROR)
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/ErrorsNative.kt]: Not "incompatible". More like "inapplicable".

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt (hunk 1 above)
Comment (LepilkinaElena) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt]: It should be another type of error as I understand

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.jetbrains.kotlin.resolve.checkers.DeclarationCheckerContext
+
+object NativeSharedImmutableChecker : DeclarationChecker {
+    private val sharedImmutableFqName = FqName("kotlin.native.concurrent.SharedImmutable")
+
+    override fun check(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        check(declaration, descriptor, context, ErrorsNative.INAPPLICABLE_SHARED_IMMUTABLE_VAR) {
+            descriptor !is VariableDescriptor || !descriptor.isVar
+        }
+        check(declaration, descriptor, context, ErrorsNative.INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL) {
```
Comment (olonho) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt]: We shall somehow check that property actually has backing store, as with delegates we cannot say for sure if it's wrong or not.

Comment (olonho) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt]: Although, putting `@SharedImmutable` on property with custom getter may be a mistake anyway, hard to tell.

Comment (LepilkinaElena) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt]: > We shall somehow check that property actually has backing store

Backing store can be only for var, can't it?

File: compiler/testData/diagnostics/nativeTests/sharedImmutable.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+var point2 = Point(2.0, 2.0)
+
+class Date(<!INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL!>@SharedImmutable<!> val month: Int, <!INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL, INAPPLICABLE_SHARED_IMMUTABLE_VAR!>@SharedImmutable<!> var day:Int)
+class Person(val name: String) {
+    <!INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL, INAPPLICABLE_SHARED_IMMUTABLE_VAR!>@SharedImmutable<!>
+    var surname: String? = null
+}
+
+class Figure {
+    @SharedImmutable
```
Comment (SvyatoslavScherbina) [inline compiler/testData/diagnostics/nativeTests/sharedImmutable.kt]: I guess `@SharedImmutable` on computed properties should be an error too.

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt [hunk 6]
```diff
@@ -20,6 +20,8 @@ private val DIAGNOSTIC_FACTORY_TO_RENDERER by lazy {
             ErrorsNative.INCOMPATIBLE_THROWS_INHERITED, "Member inherits different @Throws filters from {0}",
             Renderers.commaSeparated(Renderers.NAME)
         )
+        put(ErrorsNative.INAPPLICABLE_SHARED_IMMUTABLE_VAR, "@SharedImmutable isn't applicable to var")
+        put(ErrorsNative.INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL, "@SharedImmutable is applicable only to top level entities")
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt]: > entities

declarations

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            DescriptorUtils.isTopLevelDeclaration(descriptor) ||
+                    (descriptor is PropertyDescriptor && declaration is KtProperty && (declaration.delegate != null ||
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeSharedImmutableChecker.kt]: No tests for delegates.

File: compiler/testData/diagnostics/nativeTests/sharedImmutable.kt [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+<!INAPPLICABLE_SHARED_IMMUTABLE_PROPERTY!>@SharedImmutable<!>
+var age = 20
+get() {
+    println("Age is: $field")
+    return field
+}
+set(value) {
+    println(value)
+}
```
Comment (SvyatoslavScherbina) [inline compiler/testData/diagnostics/nativeTests/sharedImmutable.kt]: Formatting?

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt [hunk 9]
```diff
@@ -20,6 +20,8 @@ private val DIAGNOSTIC_FACTORY_TO_RENDERER by lazy {
             ErrorsNative.INCOMPATIBLE_THROWS_INHERITED, "Member inherits different @Throws filters from {0}",
             Renderers.commaSeparated(Renderers.NAME)
         )
+        put(ErrorsNative.INAPPLICABLE_SHARED_IMMUTABLE_PROPERTY, "@SharedImmutable isn't applicable to current property")
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt]: 1. Kinda unclear. The message should list exact requirements for `@SharedImmutable`.
2. "current" doesn't fit here. Maybe "this"?

File: compiler/testData/diagnostics/nativeTests/sharedImmutable.kt [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+var globalAge = 30
+<!INAPPLICABLE_SHARED_IMMUTABLE_PROPERTY!>@SharedImmutable<!>
+var age1 = 20
+    get() {
+        println("Age is: $field")
+        return field
+    }
+    set(value) {
+        globalAge = value
+    }
```
Comment (SvyatoslavScherbina) [inline compiler/testData/diagnostics/nativeTests/sharedImmutable.kt]: Is there a test like `age1` but without setter?
I.e.
```kotlin
@SharedImmutable
val age1 = 20
    get() {
        println("Age is: $field")
        return field
    }
```

File: compiler/testData/diagnostics/nativeTests/sharedImmutable.kt [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+<!INAPPLICABLE_SHARED_IMMUTABLE_PROPERTY!>@SharedImmutable<!>
+var point4: Point
+    get() = point2
+    set(value) {
+        println(value)
+    }
+
+@ThreadLocal
+var point0 = Point(2.0, 2.0)
```
Comment (SvyatoslavScherbina) [inline compiler/testData/diagnostics/nativeTests/sharedImmutable.kt]: Also `@SharedImmutable @ThreadLocal` should probably be forbidden.

Comment (LepilkinaElena) [reply compiler/testData/diagnostics/nativeTests/sharedImmutable.kt]: Decided to do separately later.

**Pull request merged**